### PR TITLE
Make invalid Url values like (Url Http _ :: Url 'Https) impossible using GADTs 

### PR DESCRIPTION
**Pull request by Aver1y**
Note that this branch is based off of my other PR #84. With that merged this diff would be somewhat more readable. I didn't base the branch off of origin/master, because there is some interdependence between the changes.

- Introduces singleton type `SScheme scheme` with constructors `SHttp 'Http` and `SHttps 'Https`
- Changes `Url` to `data Url scheme = Url (SScheme scheme) ...`

### CODE REVIEW

**Comment by mrkkrp** [comment]
Will review after #84 is merged.

**Comment by mrkkrp** [comment]
We can now rebase this on top of current master.

**Comment by Aver1y** [comment]
Done

File: Network/HTTP/Req.hs [hunk 1]
```diff
@@ -3,18 +3,19 @@
 {-# LANGUAGE DataKinds #-}
 {-# LANGUAGE DeriveDataTypeable #-}
 {-# LANGUAGE DeriveGeneric #-}
-{-# LANGUAGE DeriveLift #-}
 {-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
 {-# LANGUAGE GeneralizedNewtypeDeriving #-}
-{-# LANGUAGE KindSignatures #-}
```
Comment (Aver1y) [inline Network/HTTP/Req.hs:9]: Went through all the extensions I thought I added and tested if they were necessary. This wasn't, but I also actually didn't add it, so it's strictly speaking unrelated to the rest of the PR.

**Comment by mrkkrp** [comment]
How does this change improve the end-user experience?

**Comment by Aver1y** [comment]
The end user experience is mostly unaffected, other than a smaller chance of bugs in the future. Here are some very minor changes:

- You can't create an invalid value using `coerce`. This could also be achieved using a type role annotation.
- You can't create an invalid value using the `Data` or `Generic` instances.

As for preventing bugs, I almost didn't notice that `lift` forgot about the scheme and that had to add a type annotation. So that's one hypothetical bug it could have prevented. Though if you decide that this complicates the code too much to be worth it, I get that too.

One alternative I thought about to not add `SScheme` to `Url` and instead just require a `Typeable` instance. `SScheme` would still be useful for pattern matching, but we could cut out the `Data` instance for it which is maybe the ugliest part of this code.

```haskell
newtype Url scheme = Typeable scheme => Url (NonEmpty ...)
```

We'd loose the Generic instance for `Url` compared to this approach, however compared to the original state I'd say the Generic instance was kind of a lie anyway.

You can see that in approach in #86.

**Comment by mrkkrp** [comment]
I think this is one of those cases when the marginal improvements in type safety are not worth the added complexity. I especially do not like that `Data` is defined manually.

Let's add that type role annotation, and I'd also want to have at least one test for the quasi-quoter that you added.

**Comment by Aver1y** [comment]
Ok

**Comment by mrkkrp** [comment]
@Aver1y I added a role annotation for `Url`: https://github.com/mrkkrp/req/commit/f059679686512595c1f52c0dfca4bebb8251113b.

**Pull request closed without merging**